### PR TITLE
chore(cvsb-18677): add pagination to getActivities

### PR DESCRIPTION
**Pull request by m-ujaffer**
##  Pagination to activities service

Add pagination to activities
[link to ticket number](https://jira.dvsacloud.uk/browse/CVSB-18677)

## Checklist

- [x] Code has been tested manually
- [x] PR title includes the JIRA ticket number
- [x] Branch is rebased against the latest develop
- [ ] Squashed commit contains the JIRA ticket number

### CODE REVIEW

File: src/services/DynamoDBService.ts [hunk 1]
```diff
@@ -54,46 +52,53 @@ export class DynamoDBService {
             .promise();
     }
 
+    /**
+     * Scan the entire table and retrieve all data
+     * @returns Promise<PromiseResult<DocumentClient.ScanOutput, AWSError>>
+     */
+         public getActivities(filterParams: any): Promise<IActivity[]> {
```
Comment (bristol-mike-jones) [inline src/services/DynamoDBService.ts:59]: Why is this `any`? won't it always be `IActivityParams` ?

**Comment by bristol-mike-jones** [comment]
We should be adding Integration tests for this too, so that we confirm the pubnlic end points are surfacing the new behaviour

File: package.json [hunk 2]
```diff
@@ -12,7 +12,7 @@
     "build": "node_modules/typescript/bin/tsc --rootDir ./ --outDir .build --sourceMap false && npm run build:copy",
     "build:dev": "node_modules/typescript/bin/tsc && npm run build:copy",
     "build:copy": "find src -type f \\( -name \"*.yml\" -or -name \"*.json\" \\) | cpio -pdm .build && find tests -type f \\( -name \"*.yml\" -or -name \"*.json\" \\)  | cpio -pdm .build",
-    "test": "BRANCH=local SLS_DEBUG=* AWS_XRAY_CONTEXT_MISSING=LOG_ERROR jest --testMatch=\"**/*.unitTest.ts\"",
+    "test": "BRANCH=local SLS_DEBUG=* jest --testMatch=\"**/*.unitTest.ts\" --runInBand",
```
Comment (bristol-mike-jones) [inline package.json:15]: How come we removed the SLS_DEBUG ? I could never get the Int test to work without it. Also these are gonna run slower if we have them `runinband` why add that?

**Review by bristol-mike-jones — CHANGES_REQUESTED** [review]
Some bits and bobs

File: src/services/GetActivitiesService.ts [hunk 3]
```diff
@@ -22,75 +21,39 @@ export class GetActivityService {
...
-    public async getActivitiesForCleanup(fromStartTime: string): Promise<IActivity[]> {
-        const activityDay = moment().format("YYYY-MM-DD");
-        const activities = await this.dbClient.getActivitiesWhereStartTimeGreaterThan(activityDay, fromStartTime);
-        if (!activities.Count) {
-            return Promise.reject({statusCode: 404, message: HTTPRESPONSE.NO_RESOURCES});
-        }
-        return activities.Items as IActivity[];
-    }
+        try {
+            const {fromStartTime, toStartTime, activityType} = params && params;
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/services/GetActivitiesService.ts:27]: This is to short circuit params. If params is not false then we can destructure it. This is alternate of:

```
if(params) {
const {fromStartTime, toStartTime, activityType} = params;
}
```

File: src/utils/Filters.ts [hunk 4]
```diff
@@ -50,20 +51,17 @@ export class ActivityFilters {
...
         };
         activities.sort(sortDateDesc);
-        if (activities.length <= 10) {
-            return activities;
-        } else {
-            activities.length = 10;
-            return activities;
-        }
+        const {length} = activities;
+        activities.length = length <= 10? length: 10;
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/utils/Filters.ts:64]: This is as per existing business logic.

File: serverless.yml [hunk 5]
```diff
@@ -104,6 +108,17 @@ resources:
         ProvisionedThroughput:
           ReadCapacityUnits: 1
           WriteCapacityUnits: 1
+          - IndexName: ActivityTypeIndex
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply serverless.yml:111]: removed

File: src/models/Activity.ts [hunk 6]
```diff
@@ -22,3 +22,12 @@ export interface IActivity {
     notes?: string;
     activityDay?: string;
 }
+
+export interface IActivityParams {
+    fromStartTime: string;
+    toStartTime?: string;
+    activityType?: string;
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/models/Activity.ts:29]: removed

File: src/functions/getActivity.ts [hunk 7]
```diff
@@ -2,11 +2,16 @@ import { GetActivityService } from "./../services/GetActivitiesService";
...
 
 
 const getActivity: Handler = async (event: any, context?: Context): Promise<any> => {
+    if (!(event && event.queryStringParameters)) {
+        return new HTTPResponse(400, HTTPRESPONSE.BAD_REQUEST);
+    }
+
     const activityService = new GetActivityService(new DynamoDBService());
-    return activityService.getActivities(event)
+    return activityService.getActivities(event.queryStringParameters)
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/functions/getActivity.ts:14]: the logic is updated and now `queryStringParameters` is destructured and passed as params to the getActivities function.

File: src/services/DynamoDBService.ts (hunk 1 above)
Comment (m-ujaffer) [reply src/services/DynamoDBService.ts:59]: updated to IActivityParams

File: src/services/DynamoDBService.ts [hunk 8]
```diff
@@ -54,46 +52,53 @@ export class DynamoDBService {
             .promise();
     }
 
+    /**
+     * Scan the entire table and retrieve all data
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/services/DynamoDBService.ts:56]: Updated the documentation

File: src/services/DynamoDBService.ts [hunk 9]
```diff
@@ -54,46 +52,53 @@ export class DynamoDBService {
...
+                ExpressionAttributeValues: {
+                    ...expressionAttributeValues
+                }
+            };
+            return this.queryAllData(params);
+        }
+
     /**
      * Retrieves the ongoing activity for a given staffId
      * @param staffId - staff id for which to retrieve activity
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/services/DynamoDBService.ts:77]: updated documentation

File: src/services/GetActivitiesService.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/services/GetActivitiesService.ts:27]: logic is removed as it is no longer required.

File: package.json (hunk 2 above)
Comment (m-ujaffer) [reply package.json:15]: package.json is updated now after the rebase.

File: src/services/GetActivitiesService.ts [hunk 10]
```diff
@@ -22,75 +21,39 @@ export class GetActivityService {
...
+            const data = await this.dbClient.getActivities(params);
+            if (!(data && data.length))  {
+                throw new HTTPResponse(404, HTTPRESPONSE.NO_RESOURCES);
+            }
+            const ActivityFilter: ActivityFilters = new ActivityFilters();
+            const result = ActivityFilter.returnOrderedActivities(data);
+            return result;
+        } catch (error) {
+            if (error instanceof HTTPResponse) {
+                throw new HTTPResponse(error.statusCode, error.body);
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/services/GetActivitiesService.ts:41]: Log added

File: src/services/GetActivitiesService.ts [hunk 11]
```diff
@@ -22,75 +21,39 @@ export class GetActivityService {
...
             }
         }
     }
+
+    // /**
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/services/GetActivitiesService.ts:46]: removed

File: src/services/GetActivitiesService.ts [hunk 12]
```diff
@@ -22,75 +21,39 @@ export class GetActivityService {
...
+    //  * Get activities of type visit based on startTime from Dynamodb
+    //  * @param fromStartTime - query param used for range key
+    //  * @returns Promise - Array of activities of type visit filtered based on the given params
+    //  */
+    // public async getActivitiesForCleanup(params: any): Promise<IActivity[]> {
+    //     // const activityDay = moment().format("YYYY-MM-DD");
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/services/GetActivitiesService.ts:52]: removed

File: src/functions/getActivity.ts [hunk 13]
```diff
@@ -1,18 +1,29 @@
...
-  return activityService
-    .getActivities(event)
-    .then((data: any) => {
-      return new HTTPResponse(200, data);
-    })
-    .catch((error: HTTPResponse) => {
-      return error;
-    });
+  const { fromStartTime, toStartTime, activityType, testStationPNumber, testerStaffId } =
+    event.queryStringParameters && event.queryStringParameters;
```
Comment (bristol-mike-jones) [inline src/functions/getActivity.ts:15]: Why is event.queryStringparameters in there twice?

Comment (m-ujaffer) [reply src/functions/getActivity.ts:15]: It is used to short-circuit and assign `queryStringParameters`. We can remove this though as we are checking `queryStringParameters` existence on line 9.

**Review by bristol-mike-jones — DISMISSED** [review]

**Review by JonLazariniDD — DISMISSED** [review]

**Review by JonLazariniDD — DISMISSED** [review]

**Review by m-coslett — DISMISSED** [review]

**Review by m-coslett — APPROVED** [review]

**Review by Daniel-Searle — APPROVED** [review]

**Pull request merged**